provider/cldr/src/transform/dates/patterns.rs

```diff
@@ -182,9 +183,96 @@ impl From<&cldr_json::DateTimeFormats> for gregory::patterns::DateTimeFormatsV1
+        let pattern_long = pattern::Pattern::from_bytes(pattern_str_long).unwrap();
+        let pattern_medium = pattern::Pattern::from_bytes(pattern_str_medium).unwrap();
+        let pattern_short = pattern::Pattern::from_bytes(pattern_str_short).unwrap();
+
+        let mut preferred_hour_cycle: Option<CoarseHourCycle> = None;
+        for hour_cycle in [
+            pattern::transform_hour_cycle::determine_coarse_hour_cycle(&pattern_full),
+            pattern::transform_hour_cycle::determine_coarse_hour_cycle(&pattern_long),
+            pattern::transform_hour_cycle::determine_coarse_hour_cycle(&pattern_medium),
+            pattern::transform_hour_cycle::determine_coarse_hour_cycle(&pattern_short),
+        ]
+        .iter()
+        {
+            if let Some(hour_cycle) = hour_cycle {
+                if let Some(preferred_hour_cycle) = preferred_hour_cycle {
+                    assert_eq!(
+                        *hour_cycle, preferred_hour_cycle,
+                        "A locale contained a mix of flexible hour cycle types"
+                    );
+                } else {
+                    preferred_hour_cycle = Some(*hour_cycle);
+                }
+            }
+        }
+
+        let preferred_hour_cycle =
+            preferred_hour_cycle.expect("Could not find a preferred hour cycle.");
+        let alt_hour_cycle = if preferred_hour_cycle == CoarseHourCycle::H11H12 {
+            CoarseHourCycle::H23H24
+        } else {
+            CoarseHourCycle::H11H12
+        };
+
+        let (time_h11_h12, time_h23_h24) = {
+            let time = (&other.calendars.gregorian.time_formats).into();
+            let alt_time = gregory::patterns::LengthPatternsV1 {
+                full: pattern::transform_hour_cycle::apply_coarse_hour_cycle(
+                    &date_time_formats_v1,
+                    &pattern_str_full,
+                    pattern_full,
+                    alt_hour_cycle,
+                )
+                .unwrap()
+                .into(),
+                long: pattern::transform_hour_cycle::apply_coarse_hour_cycle(
+                    &date_time_formats_v1,
+                    &pattern_str_long,
+                    pattern_long,
+                    alt_hour_cycle,
+                )
+                .unwrap()
+                .into(),
+                medium: pattern::transform_hour_cycle::apply_coarse_hour_cycle(
+                    &date_time_formats_v1,
+                    &pattern_str_medium,
+                    pattern_medium,
+                    alt_hour_cycle,
+                )
+                .unwrap()
+                .into(),
+                short: pattern::transform_hour_cycle::apply_coarse_hour_cycle(
+                    &date_time_formats_v1,
+                    &pattern_str_short,
+                    pattern_short,
+                    alt_hour_cycle,
+                )
+                .unwrap()
+                .into(),
+            };
```

Comment:
`nit (n-b)`: I'm trying to minimize the amount of `unwraps` in our code. Can you use `expect` here? It reads a bit better when errors out.

Reply:
I did these as wall all other new ones I added.